### PR TITLE
implement find prefix

### LINKED ISSUE #94: A new variant for path.Tree.Find() is needed
Basepaths will need to be stripped from the request path. When matching for basepaths, the result should included the matched part of the actual request path for correct trimming. Otherwise, any basepath that include `*` will not be working properly.

### PR DESCRIPTION
**Pull request by rockspore**
fix #94

### CODE REVIEW

File: path/tree.go [hunk 1]
```diff
@@ -123,6 +130,41 @@ func (t *tree) Find(path []string, index int) interface{} {
 	return nil
 }
 
+// FindPrefix searches the tree given the subpath starting from given index in
+// the path array.
+// It returns the value stored at the last matched node as well as the length of
+// the matched path segments.
+// Double wildcard will not be matched.
```
Comment (theganyo) [inline path/tree.go:137]: Shouldn't it still match everything it can?

Comment (rockspore) [reply path/tree.go:137]: We don't have `**` in our basepath. Why should we support it?

Comment (rockspore) [reply path/tree.go:137]: Also, I realize we are not following the "best match" behavior in the Find() method. For `a/b/c` would win over `a/**/c/**/f` for a path `a/b/d/c/e/f` because the former is more explicit. It's very hard to modify Find() at this point as it is aiming at completely match a given path.

Comment (rockspore) [reply path/tree.go:137]: So I would prefer we stay simple for now and make sure the basepath selection works as how Apigee has been doing.

Comment (theganyo) [reply path/tree.go:137]: I don't suppose it matters. I was just hoping that the core code could be potentially refactored and reused instead of copied. Perhaps this is simpler, though.

File: path/tree.go [hunk 2]
```diff
@@ -139,16 +181,17 @@ func (t *tree) FindAndExtract(path []string, index int) (val interface{}, varMap
...
 	}
+
 	name := path[index]
 	if name == "" { // skip empty
 		return t.findNode(path, 1+index, matchCount, varMap)
 	}
 
 	// non-wildcard match
 	if child, ok := t.children[name]; ok {
-		found = child.(*tree)
```
Comment (theganyo) [inline path/tree.go:151]: Why was this removed?

File: path/tree.go (hunk 1 above)
Comment (rockspore) [reply path/tree.go]: I added the double wildcard, which is easier than I expected, because under "best match", it's very hard to reach a doulble wildcard.

File: path/tree.go (hunk 2 above)
Comment (rockspore) [reply path/tree.go:151]: This doesn't have any effect. No test was broken without it. Can you name some example where this is needed?

**Comment by rockspore** [comment]
Added comment to `FindPrefix()` to better explain how it works.

File: path/tree.go (hunk 2 above)
Comment (theganyo) [reply path/tree.go:151]: Hmm. I don't think it will affect the outcome, but I believe removing might break the optimization the follows it.

Comment (rockspore) [reply path/tree.go:151]: I investigated a bit. So with this, it will work like a prefix matching for non-wildcard cases - `/a/b/c/d will return a `/a/b/c` given it is the "best" match. However, in any wildcard cases it breaks in the sense that `/b/c/d` will still return nil even if there is a node `/*`. Adding this same line to the single-wildcard matching condition also won't work for this because it will break other matching behaviors, e.g. `for: a/b/c/d/e/f, want: a/**/c/**/f, got: *`.

I don't see any other optimization this has. `found` is a local value and we don't do any early-termination based on that. Since we are having a separate prefix matching method, I don't think we should keep it.

Comment (theganyo) [reply path/tree.go:151]: What I'm hearing is inconsistent. You said it didn't have an effect, but after investigation, it does. We must be missing a test. Also, the variable is "local" to the function, but certainly not the block... it could be returned unchanged after this block in the `if foundMatchCount >= len(path)` check and also potentially after that.

**Review by theganyo — APPROVED** [review]
